sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -37,6 +37,13 @@ abstract class Expression extends TreeNode[Expression] {
    *  - A [[Cast]] or [[UnaryMinus]] is foldable if its child is foldable
    */
   def foldable: Boolean = false
+
+  /**
+   * Returns true when an expressions always return the same result for a specific set of
+   * input values.
+   */
+  // TODO: Need to well define what are explicit input values and implicit input values.
+  def deterministic: Boolean = true
```

Comment:
I think it depends on how you interpret this. The definition here is whether this expression itself is deterministic or not (i.e. assuming fixed input, is the output deterministic?)